Xmlconsole

barun511 opened the pull request:
In Progress: Teaser Task for Gsoc - Sending raw xml via a "Send XML" button.

Steps:

Done
1. Create SendXML button.
2. Create a QTextEdit instance that pops up on button click.
3. Create a Class (QtXMLSenderWidget) that holds a textedit and a button to send the raw XML.

Not Done
4. On Button press, send XML in QtXMLSenderWidget.
5. Check for validity of XML before sending (optional)

Screenshots of work so far attached (only User Interface).

![image](https://cloud.githubusercontent.com/assets/12569516/24295101/49a7a6a0-10bf-11e7-99a1-51cfce355b6a.png)

![image](https://cloud.githubusercontent.com/assets/12569516/24295114/541bd908-10bf-11e7-9b71-9187d35ebfdf.png)

tfar commented:
I think an input field in the original dialog would be preferred over opening an addition window to input the XML. Other than that it look great.

tfar commented on Swift/QtUI/QtXMLConsoleWidget.cpp:
```diff
@@ -46,17 +46,26 @@ QtXMLConsoleWidget::QtXMLConsoleWidget() {
 }
 
+void QtXMLConsoleWidget::showWindow() {
+    QtXMLSenderWidget* XMLWindow = new QtXMLSenderWidget();
```
You can use `auto` here on the left-hand side.

tfar commented on Swift/QtUI/QtXMLSenderWidget.cpp:
```diff
@@ -0,0 +1,38 @@
+ * Copyright (c) 2010-2016 Isode Limited.
+ * All rights reserved.
+ * See the COPYING file for more information.
+ */
```
If this file is completely new and written by yourself, it should have a copyright header with your name on it. See for example how another student has done it in their PR https://github.com/swift/swift/pull/46/files .

tfar commented on Swift/QtUI/QtXMLSenderWidget.cpp:
```diff
@@ -0,0 +1,38 @@
+#include <Swift/QtUI/QtXMLSenderWidget.h>
+
+#include <string>
+#include <QString>
```
We tend to group our includes, so ideally the QString should be grouped with other Qt includes. There is also a script doing this for you, see the python script https://github.com/swift/swift/blob/master/BuildTools/FixIncludes.py .

barun511 commented:
Changed to input field in original dialog, and changed copyright. 

I'm having trouble figuring out how to send the outgoing XML, any tips on what code I should be looking at?

tfar commented:
The MainController should have a Swift::Client instance. It has sendMessage() and sendPresence) method. For sending IQs you can also get the IQRouter from that Swift::Client instance. They are all defined in Swift::CoreClient https://github.com/swift/swift/blob/master/Swiften/Client/CoreClient.h .

tfar commented on Swift/QtUI/QtXMLConsoleWidget.cpp:
```diff
@@ -40,30 +45,34 @@ QtXMLConsoleWidget::QtXMLConsoleWidget() {
-void QtXMLConsoleWidget::showWindow() {
-    QtXMLSenderWidget* XMLWindow = new QtXMLSenderWidget();
-    XMLWindow->show();
+void QtXMLConsoleWidget::showWindow(bool showXMLSender) {
```
While this way is totally fine, you can also use C++11 lambdas as Qt slots. This has the widget definition and the handler for the widget close to each other in the code and doesn't require an additional method to be added to the class. See https://github.com/swift/swift/blob/master/Swift/QtUI/QtBlockListEditorWindow.cpp#L113 for an example.

barun511 replied:
Also made this change :)

tfar commented on Swift/QtUI/QtXMLConsoleWidget.cpp:
```diff
@@ -40,30 +45,34 @@ QtXMLConsoleWidget::QtXMLConsoleWidget() {
+    else
+    {
+        XMLWindow->hide();
+    }
```
Instead of this if you could also pass a boolean flag to `QWidget::setVisible` method, see http://doc.qt.io/qt-5/qwidget.html#visible-prop .

barun511 replied:
Made change. Will push when I get back to college tomorrow :)

tfar commented on Swift/QtUI/QtXMLConsoleWidget.h:
```diff
@@ -42,5 +42,7 @@ namespace Swift {
         private:
             QTextEdit* textEdit;
             QCheckBox* enabled;
+            QCheckBox* debugenabled;
```
Ideally all pointers are initialised with `nullptr` if you don't directly assign a correct value to them at declaration time.

barun511 replied:
I'm not entirely sure what you mean, could I have a bit more detail?

tfar replied:
The value of these members is undefined until you assign a value to them. 
You can write `QCheckBox* debugEnabled = nullptr` in the class definition in the header file to do so. Furthermore it should be `debugEnabled` and not `debugenabled`, as we camelCase for member variables. See the Naming section in https://github.com/swift/swift/wiki/Coding-Style .

barun511 commented:
4. On button press, Send XML <-- Done

tfar commented on Swift/Controllers/UIInterfaces/XMLConsoleWidget.h:
```diff
@@ -7,6 +7,7 @@
 #pragma once
 
 #include <Swiften/Base/SafeByteArray.h>
+#include <boost/signals2.hpp>
```
In Swift we aim to keep our includes sorted. There's a handy script for this which aims to do the job for most cases, https://github.com/swift/swift/blob/master/BuildTools/FixIncludes.py . Simply run it like `./BuildTools/FixIncludes.py -i path_to_cpp_or_h_file.cpp`.

tfar commented on Swift/Controllers/XMLConsoleController.cpp:
```diff
@@ -41,5 +45,11 @@ void XMLConsoleController::handleDataWritten(const SafeByteArray& data) {
+void XMLConsoleController::sendXML(std::string data){
+    client_->sendData(data);
 }
+}
```
There is a newline missing at the end of the file. While it's not required for C++11, currently all files in Swift have newlines at the end of the source files and our CI enforces this.

tfar commented on Swift/Controllers/XMLConsoleController.cpp:
```diff
@@ -41,5 +45,11 @@ void XMLConsoleController::handleDataWritten(const SafeByteArray& data) {
+    this->client_ = client_;
+}
+void XMLConsoleController::sendXML(std::string data){
+    client_->sendData(data);
```
This should probably check if `client_` is actually set, not? If the developer forgets to call `setClient(…)` pressing the send XML button would cause a crash I think.

tfar commented on Swift/Controllers/XMLConsoleController.h:
```diff
@@ -12,13 +12,15 @@
 #include <boost/signals2.hpp>
 
 #include <Swiften/Base/SafeByteArray.h>
+#include <Swiften/Client/Client.h>
```
If the predeclaration below is sufficient, then there's no need to include `Client.h` here, right? If not, the predeclaration is not needed anymore.

tfar commented on Swift/QtUI/QtXMLSenderWidget.cpp:
```diff
@@ -0,0 +1,43 @@
+ * Copyright (c) 2017 Barun Parruck.
+ * All rights reserved.
+ * See the COPYING file for more information.
+ */
```
Your copyright header should match the license you aim to submit the patch under. See for example https://github.com/swift/swift/commit/cc873b3f00db4cd0a778bc2ec04f8748d70a92f9#diff-0856e1d938fb0b1106ce0267bbb319b0R1 .

tfar commented on Swift/QtUI/QtXMLSenderWidget.h:
```diff
@@ -0,0 +1,28 @@
+/*
```
Is an additional widget really needed in this case, or wouldn't it be more straightforward to just put QTextEdit in the existing QtXMLConsoleWidget class.  I think that would make the code a bit cleaner.

barun511 commented:
I've made the necessary changes :) Anything else?

tfar commented on Swift/Controllers/XMLConsoleController.cpp:
```diff
@@ -41,5 +43,13 @@ void XMLConsoleController::handleDataWritten(const SafeByteArray& data) {
+{
+    this->client_ = client_;
+}
+void XMLConsoleController::sendXML(std::string data){
```
Please add a newline between method definitions.

tfar commented on Swift/Controllers/XMLConsoleController.h:
```diff
@@ -28,12 +29,14 @@ namespace Swift {
         private:
             void handleUIEvent(std::shared_ptr<UIEvent> event);
-
+            void sendXML(std::string data);
```
The parameter should have the type `const std::string&` to avoid a copy.

Kev commented:
Hi there, I've not looked at this yet, is it possible to get a screenshot of how the current input dialog looks, and what the button in the XML window looks like, please?

barun511 commented:
![image](https://cloud.githubusercontent.com/assets/12569516/24664364/197f4200-1978-11e7-8402-d4a7773d5d82.png)

Kev commented:
Great, thanks. Some comments on that:
Sending XML is a very rare thing to want to do, so we shouldn't clutter the debug dialog with it directly, but hide it behind a button (or is that what manually enter XML does? If so, it should probably have a better name, like 'show XML entry tools' or somesuch.

It looks like at the moment it's only letting you put in a block of XML on its own, and doesn't have any mechanisms for choosing what stanza type you want to be sending, what the values of to/from/type/id should be?

barun511 commented:
(That is what manually enter XML does, I'll change the name). 

No mechanisms for now, just the raw stuff.

Kev commented:
Thanks. It's probably not worth us spending much time reviewing the patch until the functionality is there, is it?

barun511 commented:
Well, you can still send raw XML (or even non-XML), just not specific types yet. I'm more than happy to also hold off on reviewing until I add more functionality (though I'm really grateful to Tobias for reviewing, it's helped me fix a number of issues I wouldn't have been aware of).
